### PR TITLE
Fix trace data deserialization and move modules

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/11772?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11772/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11772
```

</p>
</details>

### What changes are proposed in this pull request?

Fixing an error during `get_trace()` call in Databricks, caused by a bug introduced in #11736.
```
     23 @classmethod
     24 def from_dict(cls, d):
---> 25     return cls(spans=[Span(**span) for span in d["spans"]])

TypeError: Span.__init__() got an unexpected keyword argument 'name'
```

This PR fixes the error and enhancing test for `get_trace()` API.

**Note**
In addition to the main change, as a follow-up of the above PR, the `Span`/`LiveSpan` classes are moved from `wrapper.py` to `entities/` directory. The majority of code diff in this PR is due to this change, and you can exclude those noise by looking at [the second commit](https://github.com/mlflow/mlflow/commit/d4743cac47ddb2a8f0d78847f0dc92725147573e) only.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

```
import mlflow

client = mlflow.MlflowClient()
trace = client.get_trace(request_id="tr-550fad426a4e4bbeb9ba7022b025e9df")
```

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

**Review by harupy — APPROVED** [review]
LGTM!

File: mlflow/entities/span.py [hunk 1]
```diff
@@ -14,3 +40,434 @@ class SpanType:
...
+                INVALID_PARAMETER_VALUE,
+            )
+
+        self._span = otel_span
+        # Since the span is immutable, we can cache the attributes to avoid the redundant
+        # deserialization of the attribute values.
+        self._attributes = _CachedSpanAttributesRegistry(otel_span)
+
+    @property
+    @lru_cache(maxsize=1)
```
Comment (harupy) [inline mlflow/entities/span.py:69]: Can we use https://docs.python.org/3/library/functools.html#functools.cached_property?

Comment (B-Step62) [reply mlflow/entities/span.py:69]: Great idea!

Comment (B-Step62) [reply mlflow/entities/span.py:69]: @harupy Hmm it seems `cached_property` has a bit unique behavior that allows direct write to the attribute like `span.request_id = xxx`. I think we shouldn't allow such mutation.

> The mechanics of [cached_property()](https://docs.python.org/3/library/functools.html#functools.cached_property) are somewhat different from [property()](https://docs.python.org/3/library/functions.html#property). A regular property blocks attribute writes unless a setter is defined. In contrast, a cached_property allows writes.

Comment (harupy) [reply mlflow/entities/span.py:69]: Got it, then never mind!

File: mlflow/entities/span.py [hunk 2]
```diff
@@ -14,3 +40,434 @@ class SpanType:
...
+            parent_id = decode_id(data["parent_id"]) if data["parent_id"] else None
+
+            otel_span = OTelReadableSpan(
+                name=data["name"],
+                context=build_otel_context(trace_id, span_id),
+                parent=build_otel_context(trace_id, parent_id) if parent_id else None,
+                start_time=data["start_time"],
+                end_time=data["end_time"],
+                attributes=data["attributes"],
+                status=SpanStatus(data["status_code"], data["status_message"]).to_otel_status(),
```
Comment (serena-ruan) [inline mlflow/entities/span.py:212]: Can we directly use OTelStatus(data["status_code"], data["status_message"])?

File: mlflow/entities/trace_data.py [hunk 3]
```diff
@@ -22,7 +22,11 @@ class TraceData:
 
     @classmethod
     def from_dict(cls, d):
-        return cls(spans=[Span(**span) for span in d["spans"]])
+        return cls(
+            request=d.get("request"),
+            response=d.get("response"),
+            spans=[Span.from_dict(span) for span in d["spans"]],
```
Comment (serena-ruan) [inline mlflow/entities/trace_data.py:28]:
```suggestion
            spans=[Span.from_dict(span) for span in d.get("spans", [])],
```

**Review by serena-ruan — APPROVED** [review]
LGTM!

File: mlflow/entities/span.py (hunk 2 above)
Comment (B-Step62) [reply mlflow/entities/span.py:212]: OTel's Status only accepts their enum ([ref](https://github.com/mlflow/mlflow/blob/tracing/mlflow/entities/span_status.py#L59)) so using our constructor is easier indeed🙂

**Comment by B-Step62** [comment]
Force merging as the failed LangChain tests are not relevant to this change and fixed by https://github.com/mlflow/mlflow/pull/11774 in the master branch.

**Pull request merged**